SNMP Docs Update (#5244)
DataDog/integrations-core: snmp/README.md

Simple Network Management Protocol (SNMP) is a standard for monitoring network-connected devices, such as routers, switches, servers, and firewalls. This check collects SNMP metrics from your network devices.

SNMP uses sysOIDs (System Object Identifiers) to uniquely identify devices, and OIDs (Object Identifiers) to uniquely identify managed objects. OIDs follow a hierarchical tree pattern: under the root is ISO, which is numbered 1. The next level is ORG and numbered 3 and so on, with each level being separated by a `.`.

A MIB (Management Information Base) acts as a translator between OIDs and human readable names, and organizes a subset of the hierarchy. Because of the way the tree is structured, most SNMP values start with the same set of objects:

> * `1.3.6.1.1`: (MIB-II) A standard that holds system information like uptime, interfaces, network stack.

Reviewer: ```suggestion
* `1.3.6.1.1`: (MIB-II) A standard that holds system information like uptime, interfaces, and network stack.
```

* `1.3.6.1.4.1`: A standard that holds vendor specific information.

## Setup

### Installation

The SNMP check is included in the [Datadog Agent][1] package. No additional installation is necessary.

### Configuration

<div class="alert alert-warning">
<b>Note</b>: The following features are in beta.
</div>

> The Datadog SNMP check scans a provided subnet, Autodiscovers network devices, and collects metrics using Datadog's sysOID mapped device profiles.

Reviewer 2: `scans a provided subnet discovering network devices` ?

> The Datadog SNMP check scans a provided subnet, auto discovers network devices, and collects metrics using Datadog's sysOID mapped device profiles.

Reviewer: ```suggestion
The Datadog SNMP check scans a provided subnet, Autodiscovers network devices, and collects metrics using Datadog's sysOID mapped device profiles.
```

let's also add a link to the Autodiscovery page on `Autodiscovers` since i believe that's the first mention of AD on this page

Reviewer 3: It's a different autodiscovery than the agent integration autodiscovery

> Edit the subnet, snmp version, and profiles in the `snmp.d/conf.yaml` file in the `conf.d/` folder at the root of your [Agent's configuration directory][2]. See the [sample snmp.d/conf.yaml][3] for all available configuration options.

Reviewer: ```suggestion
Edit the subnet, SNMP version, and profiles in the `snmp.d/conf.yaml` file in the `conf.d/` folder at the root of your [Agent's configuration directory][2]. See the [sample snmp.d/conf.yaml][3] for all available configuration options.
```

> #### Auto-Discovery

Reviewer: ```suggestion
#### Autodiscovery
```

> To use Auto-Discovery with the SNMP check:

Reviewer: ```suggestion
To use Autodiscovery with the SNMP check:
```

| `bulk_threshold`             | The number of symbols in a table that triggers a BULK request. This paramater is only relevant for SNMPv > 1.                                                                                                                                                                                                                                                                                                                                                                                                                              |
| `tags`                       | A list of global tags to add to all SNMP metrics. Read more about [tagging in Datadog][6].                                                                                                                                                                                                                                                                                                                                                                                                                                                 |

##### Sample config

```yaml
init_config:

>   @param profiles - object - optional

Reviewer 2: Need to remove that line.

Author: removed in most recent push

    f5-big-ip:
      definition_file: f5-big-ip.yaml
    router:
      definition_file: generic-router.yaml

instances:
   ## @param network_address - string - optional

>    -  

Reviewer 4: Some formatting suggestions based on the [rendered view](https://github.com/DataDog/integrations-core/blob/68ca528ab136b3bbcf9805efd6e2e556d941f41f/snmp/README.md#sample-config): I think we need to move the delimiting `-` at the top, and then align comments with their respective parameters...

```diff
-   ## @param ...
-   -
+  -
+    ## @param ...
     network_address: ...

-   ## @param ...
+    ## @param ...
     port: 161
```

Author: updated in most recent commit

>    network_address: "<NETWORK_ADDRESS>"

Reviewer 4: Isn't this missing the leading `-` used to separate instances in this array?

```yaml
instances:
  - network_address: ...
    port: 161
    ...
```

Author: Does each entry here need a `-`? Or just `network_address`?

Author: Updated for `-` before `network_address` in most recent commit. Definitely let me know if I'm missing any more `-`

Reviewer 4: `-` is used to delimit the configuration for a given instance (which represents a device, or a subnet to be discovered). AFAIU the example config describes configuration for a single auto-discovered subnet, so we should be good now, though I have pending comments on formatting. :-)

     # and use the profile if it matches.
```

If necessary, additional metrics can be defined in the instances. These metrics are collected in addition to those in the profile.

> #### Metric definition by Profile

Reviewer: ```suggestion
#### Metric definition by profile
```

Profiles can be used interchangeably, such that devices that share MIB dependencies can reuse the same profiles. For example, the Cisco c3850 profile can be used across many Cisco switches.

**Note**: Profiles require a local version of the MIBs to be referenced.

##### Generic router profile

The [generic router profile][20] collects the following metrics.

> MIBs needed for local reference: IF-MIB, IP-MIB, TCP-MIB, UDP-MIB

Reviewer: ```suggestion
MIBs needed for local reference: IF-MIB, IP-MIB, TCP-MIB, UDP-MIB.
```

| `snmp.ipSystemStatsHCInForwDatagrams`  | The number of input datagrams for which this entity was not their final IP destination and for which this entity attempted to find a route to forward them to that final destination.                                                                                     | `ipversion`              |
| `snmp.ipSystemStatsReasmReqds`         | The number of IP fragments received that needed to be reassembled at this interface.                                                                                                                                                                                      | `ipversion`              |
| `snmp.ipSystemStatsReasmOKs`           | The number of IP datagrams successfully reassembled.                                                                                                                                                                                                                      | `ipversion`              |

> | `snmp.ipSystemStatsReasmFails`         | The number of failures detected by the IP re-assembly algorithm (for whatever reason: timed out, errors, etc.).                                                                                                                                                           | `ipversion`              |

Reviewer: ```suggestion
| `snmp.ipSystemStatsReasmFails`         | The number of failures detected by the IP reassembly algorithm (for whatever reason: timed out, errors, etc.).                                                                                                                                                           | `ipversion`              |
```

| `snmp.ipIfStatsHCInForwDatagrams`      | The number of input datagrams for which this entity was not their final IP destination and for which this entity attempted to find a route to forward them to that final destination.                                                                                     | `ipversion`, `interface` |
| `snmp.ipIfStatsReasmReqds`             | The number of IP fragments received that needed to be reassembled at this interface.                                                                                                                                                                                      | `ipversion`, `interface` |
| `snmp.ipIfStatsReasmOKs`               | The number of IP datagrams successfully reassembled.                                                                                                                                                                                                                      | `ipversion`, `interface` |

> | `snmp.ipIfStatsReasmFails`             | The number of failures detected by the IP re-assembly algorithm (for whatever reason: timed out, errors, etc.).                                                                                                                                                           | `ipversion`, `interface` |

Reviewer: ```suggestion
| `snmp.ipIfStatsReasmFails`             | The number of failures detected by the IP reassembly algorithm (for whatever reason: timed out, errors, etc.).                                                                                                                                                           | `ipversion`, `interface` |
```

| `snmp.udpNoPorts`                      | The total number of received UDP datagrams for which there was no application at the destination port.                                                                                                                                                                    |                          |
| `snmp.udpInErrors`                     | The number of received UDP datagrams that could not be delivered for reasons other than the lack of an application at the destination port.                                                                                                                               |                          |
| `snmp.udpHCOutDatagrams`               | The total number of UDP datagrams sent from this entity, for devices that can transmit more than 1 million UDP datagrams per second.                                                                                                                                      |                          |

##### F5 BIG-IP profile

The [F5 BIG-IP profile][21] collects the following metrics.

> MIBs needed for local reference: F5-BIGIP-SYSTEM-MIB, IF-MIB

Reviewer: ```suggestion
MIBs needed for local reference: F5-BIGIP-SYSTEM-MIB, IF-MIB.
```

| `snmp.udpNoPorts`                        | The total number of received UDP datagrams for which there was no application at the destination port.                                                                                                                                                               |             |
| `snmp.udpInErrors`                       | The number of received UDP datagrams that could not be delivered for reasons other than the lack of an application at the destination port.                                                                                                                          |             |
| `snmp.udpHCOutDatagrams`                 | The total number of UDP datagrams sent from this entity for devices that can transmit more than 1 million UDP datagrams per second.                                                                                                                                  |             |

##### Cisco c3850 device profile

The [Cisco c3850 device profile][22] collects the following metrics.

> MIBs needed for local reference: CISCO-ENTITY-SENSOR-MIB, CISCO-ENTITY-FRU-CONTROL-MIB, CISCO-PROCESS-MIB, CISCO-IF-EXTENSION-MIB, IF-MIB, TCP-MIB, UDP-MIB

Reviewer: ```suggestion
MIBs needed for local reference: CISCO-ENTITY-SENSOR-MIB, CISCO-ENTITY-FRU-CONTROL-MIB, CISCO-PROCESS-MIB, CISCO-IF-EXTENSION-MIB, IF-MIB, TCP-MIB, UDP-MIB.
```

| `snmp.cpmCPUTotalMonIntervalValue` | The overall CPU busy percentage in the last cpmCPUMonInterval period.                                                                                                                                                  | `cpu`                      |
| `snmp.cpmCPUMemoryUsed`            | The overall CPU wide system memory currently in use.                                                                                                                                                       | `cpu`                      |
| `snmp.cpmCPUMemoryFree`            | The overall CPU wide system memory currently free.                                                                                                                                                            | `cpu`                      |
| `snmp.cieIfLastInTime`             | The elapsed time in milliseconds since the last protocol input packet was received.                                                                                                                                        | `interface`                |

> | `snmp.cieIfLastOutTime`            | The elapsed time in milliseconds since the last protocol  output packet was transmitted.                                                                                                                                   | `interface`                |

Reviewer: ```suggestion
| `snmp.cieIfLastOutTime`            | The elapsed time in milliseconds since the last protocol output packet was transmitted.                                                                                                                                   | `interface`                |
```

     # and use the profile if it matches.
```

If necessary, additional metrics can be defined in the instances. These metrics are collected in addition to those in the profile.

#### Metric definition by profile

Profiles can be used interchangeably, such that devices that share MIB dependencies can reuse the same profiles. For example, the Cisco c3850 profile can be used across many Cisco switches.

**Note**: Profiles require a local version of the MIBs to be referenced.

* [Generic router](#generic-router-profile)
* [F5 Big IP](#f5-big-ip-profile)
* [Dell iDRAC](#dell-idrac-profile)
* [Cisco Nexus](#cisco-nexus-profile)
* [Cisco c3850](#cisco-c3850-device-profile)
* [Cisco Meraki](#cisco-meraki-profile)

##### Generic router profile

The [generic router profile][7] collects the following metrics.

> MIBs needed for local reference: IF-MIB, IP-MIB, TCP-MIB, UDP-MIB.

Reviewer 4: To be confirmed by @therve, but thanks to #5397 I don't think users customers still need to have MIBs used by built-in profiles installed locally, do they?

Author: I believe you're correct-- I'll remove

| Metric                                 | Description                                                                                                                                                                                                                                                               | Tags                     |
|----------------------------------------|---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|--------------------------|

> | `snmp.ifInDiscards`                    | The number of inbound packets chosen to be discarded even though no errors had been detected to prevent them being deliverable to a higher-layer protocol.                                                                                                                | `interface`              |

Reviewer 4: These lists are huge! Are you using scripts to generate them? Is there anything we can do to make sure these stay in sync if/when we update profiles, or to make life easier for documentation?

Reviewer 5: Could we use the `metadata.csv` file for this? Adding the profile it's related to in the corresponding metrics description. This will allow updating the metric description within the DD app.

Author: I'm fine with metadata.csv, this was a temp solution just so we could ensure that we had metric definitions somewhere...  That being said, I personally am not sure of a good way to show all of the possible metrics you _could_ get from the profiles. The metadata file is usually just a giant list, and I do think that there should be some way to define what metrics you get from each profile

Author: FWIW I've been doing these by hand and they are time consuming, but I think basically just as time consuming as writing the metadata file can be

Reviewer 5: As suggested we could add the profile in the metric description with a disclaimer on top of the metric table in the doc to explain that metrics are profile dependant. I can take care of this if i have your approval @naaltman

Author: Are you thinking like as a column in the csv, a reference for the profile that the metrics come from? That works for me... There's a decent amount of overlap so that could be good I think

Reviewer 4: Any new ad-hoc `profiles` column in the CSV would need to contain a list of profiles though, as I think metrics in `metadata.csv` must be unique and there may be (and in practice there *are*) multiple profiles associated to a given metric.

The source of truth is the profiles files themselves. Ideally we'd generate `metadata.csv` from those files… and then any rendered tables from those files. (@l0k0ms I think there's some tooling in progress for metadata.csv -> Markdown tables, right?).

Not sure how industrial the profiles -> `metadata.csv` tooling needs to be, but it could be a simple utility script stored in the integration folder. Or even better: a `$ ddev meta snmp ...` command. E.g. we already have `$ ddev meta snmp translate-profile`, so we could add `$ ddev meta snmp generate-metadata`. 🤔

Reviewer 5: ha i like this approach!

There is a logic in the documentation build that indeed inlines a markdown table based on the metadata.csv file (for instance the [Apache integration](https://docs.datadoghq.com/integrations/apache/#metrics). There is also a logic in-app to display the metadata.csv file content automatically in a ["metrics" tab on an integrations tile](https://app.datadoghq.com/account/settings#integrations/apache)

Sadly, we can't change the columns of the `metadata.csv` file as there are too much dependencies for it.

The idea would be to prefix a metrics description with the associated profile:

```
| `snmp.ifInDiscards`                    | The number of inbound packets chosen to be discarded even though no errors had been detected to prevent them being deliverable to a higher-layer protocol. |  `interface`              |
```

Would become: 

```
snmp.ifInDiscards,gauge,,,, [Generic router profile] The number of inbound packets chosen to be discarded even though no errors had been detected to prevent them being deliverable to a higher-layer protocol.,0,snmp,snmp_ifindiscards
```

Two issues here:

* It's not possible to display the list of tags attached to a metric
* We need a type to associate to the metric, are they all `gauge` type ?

Reviewer 4: > There is a logic in the documentation build that indeed inlines a markdown table based on the metadata.csv

I assume that logic is executed at runtime, correct? The Apache docs page shows the rendered Markdown table, but our `README` contains a sentence:

```markdown
See [metadata.csv][9] for a list of metrics provided by this check.
```

I don't think we *need* to put the rendered table in the source of this README (providing it on the docs app via this runtime rendering mechanism is enough, as most users will look up the docs there).

So maybe there's no need for any more tooling. Let's just move stuff to `metadata.csv` as a follow-up — which I assume is what you had in mind @l0k0ms?

> We need a type to associate to the metric, are they all gauge type ?

Not all of them, no. E.g. the `cisco-nexus.yaml` profile as well as the iDRAC one contain metrics converted to `monotonic_count` metrics. (That's the case for `ifInDiscards` for example, which represents a total number of discards of input packets.)

Reviewer 5: indeed you're right, it's done at runtime. And i agree with you we don't "need" too, having it in the metadata.csv is enough.. 

Noted for the metrics type !

Let's merge this and move the metrics table in another PR for sanity

| `snmp.udpNoPorts`                      | The total number of received UDP datagrams for which there was no application at the destination port.                                                                                                                                                                    |                          |
| `snmp.udpInErrors`                     | The number of received UDP datagrams that could not be delivered for reasons other than the lack of an application at the destination port.                                                                                                                               |                          |
| `snmp.udpHCOutDatagrams`               | The total number of UDP datagrams sent from this entity, for devices that can transmit more than 1 million UDP datagrams per second.                                                                                                                                      |                          |

##### F5 BIG-IP profile

The [F5 BIG-IP profile][8] collects the following metrics.

> MIBs needed for local reference: F5-BIGIP-SYSTEM-MIB, IF-MIB.

Reviewer 2: You should be able to remove that for 7.17+ too.

Author: updated and removed the reference

     # and use the profile if it matches.
```

If necessary, additional metrics can be defined in the instances. These metrics are collected in addition to those in the profile.

#### Metric definition by profile
Profiles can be used interchangeably, such that devices that share MIB dependencies can reuse the same profiles. For example, the Cisco c3850 profile can be used across many Cisco switches.

**Note**: Profiles require a local version of the MIBs to be referenced.

* [Generic router](#generic-router-profile)

> * [F5 Big IP](#f5-big--ip-profile)

Reviewer 5: ```suggestion
* [F5 Big IP](#f5-big-ip-profile)
```

* [Dell iDRAC](#dell-idrac-profile)
* [Cisco Nexus](#cisco-nexus-profile)

> * [Cisco c3850](#cisco-c3850-profile)

Reviewer 5: ```suggestion
* [Cisco c3850](#cisco-c3850-device-profile)
```

> * [Cisco Meraki](#cisco-meraki)

Reviewer 5: ```suggestion
* [Cisco Meraki](#cisco-meraki-profile)
```

| `snmp.ifHCOutOctets`               | The total number of octets transmitted out of the interface including framing characters.                                                                                                                             | `interface`                |
| `snmp.ifHCOutUcastPkts`            | The total number of packets that higher-level protocols requested be transmitted that were not addressed to a multicast or broadcast address at this sub-layer including those discarded or not sent. | `interface`                |
| `snmp.ifHCOutMulticastPkts`        | The total number of packets that higher-level protocols requested be transmitted that were addressed to a multicast address at this sub-layer including those discarded or not sent.                  | `interface`                |
| `snmp.ifHCOutBroadcastPkts`        | The total number of packets that higher-level protocols requested be transmitted addressed to a broadcast address at this sub-layer including those discarded or not sent.                  | `interface`                |

> #### Cisco Nexus Profile

Reviewer 5: ```suggestion
##### Cisco Nexus Profile
```
